contrib/linux-agent-installer/Icinga2Agent.bash

```diff
@@ -1,76 +1,169 @@
+done
+[ -z "$errors" ] || exit 1
 
+# Make sure icinga2 is installed and running
+echo -n "check: icinga2 installed - "
+if icinga2 --version &>/dev/null; then
+  echo "OK"
+else
+  fail "install icinga2!"
+fi
 
-if [ -f "$RHEL_SYSCONFIG" ]; then
-  ICINGA2_SYSCONFIG_FILE="$RHEL_SYSCONFIG"
-elif [ -f "$DEB_SYSCONFIG" ]; then
-  ICINGA2_SYSCONFIG_FILE="$DEB_SYSCONFIG"
+# Detect osfamily
+if [ -n "$ICINGA2_OSFAMILY" ]; then
+  info "Assuming supplied osfamily $ICINGA2_OSFAMILY"
+elif [ -e /etc/sysconfig/icinga2 ] && check_command rpm && ! check_command dpkg; then
+  info "This should be a RedHat system"
+  # shellcheck disable=SC1091
+  . /etc/sysconfig/icinga2
+  ICINGA2_OSFAMILY=redhat
+elif [ -e /etc/default/icinga2 ] && check_command dpkg; then
+  info "This should be a Debian system"
+  # shellcheck disable=SC1091
+  . /etc/default/icinga2
+  ICINGA2_OSFAMILY=debian
 else
-  echo "ERROR: couldn't find your Icinga2 sysconfig file"
+  fail "Could not determine your os type!"
 fi
 
-. "$ICINGA2_SYSCONFIG_FILE"
-[ "$ICINGA2_USER" ] || fail "\$ICINGA2_USER has not been defined"
-ICINGA2_CONF_DIR=$(dirname "$ICINGA2_CONFIG_FILE")
-ICINGA2_SYSCONF_DIR=$(dirname "$ICINGA2_CONF_DIR")
-ICINGA2_INSTALL_PREFIX=$(dirname $(dirname "$DAEMON"))
-ICINGA2_CA_DIR="${ICINGA2_STATE_DIR}/lib/icinga2/ca"
-ICINGA2_SSL_DIR="${ICINGA2_CONF_DIR}/pki"
-ICINGA2_CA_PORT="5665"
+# internal defaults
+: "${ICINGA2_CONFIG_FILE:=/etc/icinga2/icinga2.conf}"
+: "${ICINGA2_CONFIGDIR:="$(dirname "$ICINGA2_CONFIG_FILE")"}"
+: "${ICINGA2_DATADIR:=/var/lib/icinga2}"
+: "${ICINGA2_SSLDIR_OLD:="${ICINGA2_CONFIGDIR}"/pki}"
+: "${ICINGA2_SSLDIR:="${ICINGA2_DATADIR}"/certs}"
+: "${ICINGA2_BIN:=icinga2}"
 
-. "${ICINGA2_INSTALL_PREFIX}/lib/icinga2/prepare-dirs" "${ICINGA2_SYSCONFIG_FILE}"
+case "$ICINGA2_OSFAMILY" in
+debian)
+  : "${ICINGA2_USER:=nagios}"
+  : "${ICINGA2_GROUP:=nagios}"
+  ;;
+redhat)
+  : "${ICINGA2_USER:=icinga}"
+  : "${ICINGA2_GROUP:=icinga}"
+  ;;
+*)
+  fail "Unknown osfamily '$ICINGA2_OSFAMILY'!"
+  ;;
+esac
 
-if ! [ -d $ICINGA2_SSL_DIR ]; then mkdir $ICINGA2_SSL_DIR; fi
-chown $ICINGA2_USER $ICINGA2_SSL_DIR
+if [ ! -d "$ICINGA2_SSLDIR" ]; then
+  mkdir "$ICINGA2_SSLDIR"
+  chown "$ICINGA2_USER.$ICINGA2_GROUP" "$ICINGA2_SSLDIR"
+fi
 
-if [ -f  "${ICINGA2_SSL_DIR}/${ICINGA2_NODENAME}.crt" ]; then
+cert_location="$ICINGA2_SSLDIR"
+cert_old_location=0
+if [ -e "$ICINGA2_SSLDIR/${ICINGA2_NODENAME}.crt" ]; then
+  : "keep default"
+elif [ -e "$ICINGA2_SSLDIR_OLD/${ICINGA2_NODENAME}.crt" ] || [ ! -d "$ICINGA2_SSLDIR_OLD" ]; then
+  cert_location="$ICINGA2_SSLDIR_OLD"
+  cert_old_location=1
+fi
+
+if [ -f  "${cert_location}/${ICINGA2_NODENAME}.crt" ]; then
   warn "ERROR: a certificate for '${ICINGA2_NODENAME}' already exists"
-  warn "Please remove ${ICINGA2_SSL_DIR}/${ICINGA2_NODENAME}.??? in case you want a"
+  warn "Please remove ${cert_location}/${ICINGA2_NODENAME}.??? in case you want a"
   warn "new certificate to be generated and signed by ${ICINGA2_CA_NODE}"
```

Comment:
If you prefer to move everything into the if/else structure that's fine for me. As there is only one large main block executing an action I'd prefer to first have all the extra-conditions, exit immediately if some precondition isn't met. And then after those run the main logic.

To me this makes it easier to read and you do not risk to run into code running when it shouldn't just because someone added some logic to the end of the script later on.

Reply:
I wanted to be able to update config files even when certificates are already present.

We could add a read action, to allow the user to continue anyways.

Reply:
Got it, that's fine.